codebuild/spec/buildspec_asan.yml

```diff
@@ -18,42 +18,86 @@ version: 0.2
 
 # Codebuild's matrix jobs have non-differentiated names so use batch-list
 # instead.
+
+# Parameter motivation
+
+# COMPILERS
+# We run asan on both gcc and clang because of different features sets for their
+# address sanitizers. Specifically there was a case where GCC was able to detect
+# a memcpy-param-overlap that Clang did not.
+
+# LIBCRYPTOS
+# awslc: happy path libcrypto for s2n-tls
+# openssl 3: s2n-tls takes different code paths for ossl3, so make sure we run 
+#            asan on it. See pr 4033 for a historical motivating example.
+# openssl 1.1.1: a widely deployed version of openssl.
+# openssl 1.0.2: the default libcrypto on AL2, and AL2 is still widely deployed.
 batch:
   build-list:
-    # awslc is the happy path libcrypto for s2n-tls
-    - identifier: awslc
+    - identifier: clang_awslc
       env:
         compute-type: BUILD_GENERAL1_LARGE
         variables:
           S2N_LIBCRYPTO: awslc
-    # s2n-tls takes different code paths for ossl3, so make sure we run asan on
-    # it. See pr 4033 for a historical motivating example.
-    - identifier: openssl_3_0
+          COMPILER: clang
+    - identifier: clang_openssl_3_0
       env:
         compute-type: BUILD_GENERAL1_LARGE
         variables:
           S2N_LIBCRYPTO: openssl-3.0
-    # openssl 1.1.1 is a widely deployed version of openssl.
-    - identifier: openssl_1_1_1
+          COMPILER: clang
+    - identifier: clang_openssl_1_1_1
       env:
         compute-type: BUILD_GENERAL1_LARGE
         variables:
           S2N_LIBCRYPTO: openssl-1.1.1
-    # openssl 1.0.2 is the default distributed on AL2, and AL2 is still widely
-    # deployed
-    - identifier: openssl_1_0_2
+          COMPILER: clang
+    - identifier: clang_openssl_1_0_2
       env:
         compute-type: BUILD_GENERAL1_LARGE
         variables:
           S2N_LIBCRYPTO: openssl-1.0.2
+          COMPILER: clang
+    - identifier: gcc_awslc
+      env:
+        compute-type: BUILD_GENERAL1_LARGE
+        variables:
+          S2N_LIBCRYPTO: awslc
+          COMPILER: gcc
+    - identifier: gcc_openssl_3_0
+      env:
+        compute-type: BUILD_GENERAL1_LARGE
+        variables:
+          S2N_LIBCRYPTO: openssl-3.0
+          COMPILER: gcc
+    - identifier: gcc_openssl_1_1_1
+      env:
+        compute-type: BUILD_GENERAL1_LARGE
+        variables:
+          S2N_LIBCRYPTO: openssl-1.1.1
+          COMPILER: gcc
+    - identifier: gcc_openssl_1_0_2
+      env:
+        compute-type: BUILD_GENERAL1_LARGE
+        variables:
+          S2N_LIBCRYPTO: openssl-1.0.2
+          COMPILER: gcc
 
 phases:
+  pre_build:
+    commands:
+      - |
+        if [ -d "third-party-src" ]; then
+          cd third-party-src;
+          ln -s /usr/local $CODEBUILD_SRC_DIR/third-party-src/test-deps;
+        fi
+      - /usr/bin/$COMPILER --version
```

Comment:
Do we expect a certain minimum gcc version to get "enough" asan coverage?

Reply:
Hmm. I do really like that idea, but I don't want to do it directly in the buildspec. Perhaps it would make sense to add a codebuild script like ```codebuild/bin/assert_compiler_version_gt $COMPILER "clang-15"```. I'll do that in a different PR to keep this one straightforward.

Reply:
Right- I'm not asking for it as part of this PR, (totally agree the buildspec probably isn't the right place) just wondering if it's written down anywhere...